Source/WebKit/UIProcess/mac/WebViewImpl.mm

```diff
@@ -5111,6 +5127,113 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
+    if (!NSEqualRanges(oldRelativeSelectedRange, relativeSelectedRange))
+        return;
+
+    NSString *paragraphContextForCandidateRequest = postLayoutData->paragraphContextForCandidateRequest;
+    auto offsetSelectionRect = postLayoutData->selectionBoundingRect;
+    offsetSelectionRect.move(0, offsetSelectionRect.height());
+
+    NSUInteger offset = absoluteSelectedRange.location - relativeSelectedRange.location;
+    NSTextCheckingResult *adjustedCandidate = [candidate resultByAdjustingRangesWithOffset:offset];
+
+    NSSpellChecker *spellChecker = [NSSpellChecker sharedSpellChecker];
+
+    if (![spellChecker respondsToSelector:@selector(showCompletionForCandidate:selectedRange:offset:inString:rect:view:completionHandler:)])
+        return;
+
+    WeakPtr weakThis { *this };
+    [spellChecker
+        showCompletionForCandidate:adjustedCandidate
+        selectedRange:absoluteSelectedRange
+        offset:offset
+        inString:paragraphContextForCandidateRequest
+        rect:offsetSelectionRect
+        view:m_view.getAutoreleased()
+        completionHandler:[weakThis](NSDictionary<NSString *, id> *resultDictionary) {
+        if (!weakThis)
+            return;
+
+        // FIXME: rdar://105809280 Adopt NSTextCheckingSoftSpaceRangeKey once it is in more builds.
+        NSValue *softSpaceRangeValue = resultDictionary[@"SoftSpaceRange"];
+        if (!softSpaceRangeValue)
+            return;
+
+        NSRange absoluteSoftSpaceRange = softSpaceRangeValue.rangeValue;
+        weakThis->selectedRangeWithCompletionHandler([weakThis, absoluteSoftSpaceRange](NSRange absoluteSelectedRange) {
+            if (!weakThis)
+                return;
+
+            auto postLayoutData = weakThis->postLayoutDataForContentEditable();
+            if (!postLayoutData)
+                return;
+
+            NSRange relativeSelectedRange = NSMakeRange(postLayoutData->candidateRequestStartPosition, postLayoutData->selectedTextLength);
+
+            if (absoluteSoftSpaceRange.location + absoluteSoftSpaceRange.length != absoluteSelectedRange.location)
+                return;
+
+            NSUInteger offset = absoluteSelectedRange.location - relativeSelectedRange.location;
+            NSRange relativeSoftSpaceRange = NSMakeRange(absoluteSoftSpaceRange.location - offset, absoluteSoftSpaceRange.length);
+
+            weakThis->m_softSpaceRange = relativeSoftSpaceRange;
+        });
+    }];
+}
+
+void WebViewImpl::showInlinePredictionsForCandidates(NSArray<NSTextCheckingResult *> *candidates)
+{
+    auto postLayoutData = postLayoutDataForContentEditable();
+    if (!postLayoutData)
+        return;
+
+    if (m_lastStringForCandidateRequest != postLayoutData->stringForCandidateRequest)
+        return;
+
+    NSSpellChecker *spellChecker = [NSSpellChecker sharedSpellChecker];
+
+    if (![spellChecker respondsToSelector:@selector(completionCandidateFromCandidates:)])
```

Comment:
Is this selector check necessary?

I think we also need a declaration in `NSSpellCheckerSPI.h` for OpenSource builds.